Installing unpublished packages results in incorrect "malformed response/registry down" warnings

**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

Example:
```
$ yarn add x-iterable-tree
yarn add v1.3.2
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
error Received malformed response from registry for "x-iterable-tree". The registry may be down.
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
```

**What is the expected behavior?**

Something that reflects that the package was unpublished and not suggesting the registry might be down.

**Please mention your node.js, yarn and operating system version.**

v1.3.2

@rally25rs rather than checking for `unpublished`, it looks like the simplest fix is to check the `versions` field of `RegistryResponse` for any valid versions and to throw an error if empty. This appears to be what npm does:

```
$ npm i x-iterable-tree

npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for x-iterable-tree
```
Let me know if this makes sense to you and I can send a PR

--- a/__tests__/registries/npm-registry.js
+++ b/__tests__/registries/npm-registry.js
@@ -724,6 +724,7 @@ describe('checkOutdated functional test', () => {
     resolveConstraints(): string {
       return '2.0.0';
     },
+    reporter: new BufferReporter({verbose: true}),
   };
 
   test('homepage URL from top level', async () => {
@@ -840,5 +841,29 @@ describe('checkOutdated functional test', () => {
       wanted: '2.0.0',
       url: 'http://package.repo.com',
     });
+  });
+
+  test('unpublished package (no versions)', async () => {
+    const testCwd = '.';
+    const {mockRequestManager, mockRegistries, mockReporter} = createMocks();
+    const npmRegistry = new NpmRegistry(testCwd, mockRegistries, mockRequestManager, mockReporter);
+
+    mockRequestManager.request = () => {
+      return {
+        'dist-tags': {
+          latest: '2.0.0',
+        },
+        versions: {},
+      };
+    };
+
+    let message;
+    try {
+      await npmRegistry.checkOutdated(mockConfig, 'left-pad', '2.0.0');
+    } catch (err) {
+      message = err.message;
+    }
+
+    expect(message).toEqual(expect.stringContaining('No valid versions'));
   });
 });

--- a/src/reporters/lang/en.js
+++ b/src/reporters/lang/en.js
@@ -325,6 +325,7 @@ const messages = {
 
   infoFail: 'Received invalid response from npm.',
   malformedRegistryResponse: 'Received malformed response from registry for $0. The registry may be down.',
+  registryNoVersions: 'No valid versions found for $0. The package may be unpublished.',
 
   cantRequestOffline: "Can't make a request in offline mode ($0)",
   requestManagerNotSetupHAR: 'RequestManager was not setup to capture HAR files',

--- a/src/resolvers/registries/npm-resolver.js
+++ b/src/resolvers/registries/npm-resolver.js
@@ -34,7 +34,11 @@ export default class NpmResolver extends RegistryResolver {
     body: RegistryResponse,
     request: ?PackageRequest,
   ): Promise<Manifest> {
-    if (!body['dist-tags']) {
+    if (body.versions && Object.keys(body.versions).length === 0) {
+      throw new MessageError(config.reporter.lang('registryNoVersions', body.name));
+    }
+
+    if (!body['dist-tags'] || !body.versions) {
       throw new MessageError(config.reporter.lang('malformedRegistryResponse', body.name));
     }
 

